DOC-5371 - two phase finality documentation and release notes (#1655)
corda/corda-docs-portal: content/en/platform/corda/4.11/enterprise/flow-state-machines.md

# Writing flows

This article explains our approach to modeling business processes and the lower level network protocols that implement
them. It explains how the platform’s flow framework is used, and takes you through the code for a simple

> 2-party asset trading flow which is included in the source.

Reviewer: ```suggestion
two-party asset trading flow which is included in the source.
```

Assuming no malicious termination, they both end the flow being in possession of a valid, signed transaction that
represents an atomic asset swap.

> Note that it is the *seller* who initiates contact with the buyer, not vice-versa as you might imagine.

Reviewer: ```suggestion
Note that it is the *seller* who initiates contact with the buyer, not vice-versa as you might expect.
```

* `acceptablePrice: Amount<Currency>` - the price that was agreed upon out of band. If the seller specifies
a price less than or equal to this, then the trade will go ahead
* `typeToBuy: Class<out OwnableState>` - the type of state that is being purchased. This is used to check that the
sell side of the flow is not trying to sell us the wrong thing, whether by accident or on purpose
* `anonymous: Boolean` - whether to generate a fresh, anonymous public key for the transaction

> Alright, so using this flow shouldn’t be too hard: in the simplest case we can just create a buyer or seller

Reviewer: ```suggestion
So, using this flow shouldn’t be too hard: in the simplest case we can just create a buyer or seller
```

* The party we wish to send the message to
* The payload being sent

> `otherSideSession.send` will serialize the payload and send it to the other party automatically.

Reviewer: ```suggestion
`otherSideSession.send` serializes the payload and sends it to the other party automatically.
```

### FinalityFlow

> On the buyer side, we use `FinalityFlow` to finalize the transaction. It will:

Reviewer: ```suggestion
On the buyer side, we use `FinalityFlow` to finalize the transaction. It does the following:
```

(And update bullets accordingly)

Author: Done

* Sends the transaction to the chosen notary and, if necessary, satisfies the notary that the transaction is valid.
* Records the transaction in the local vault, if it is relevant (that is, involves the owner of the node).
* Sends the fully signed transaction to the other participants for recording also.

> On the seller side we use `ReceiveFinalityFlow` to receive and record the finalized transaction.

Reviewer: ```suggestion
On the seller side, use `ReceiveFinalityFlow` to receive and record the finalized transaction.
```

Author: Done

{{< warning >}}
If the buyer stops before sending the finalized transaction to the seller, the buyer is left with a

> valid transaction but the seller isn’t, so they don’t get the cash! This sort of thing is not

Reviewer: ```suggestion
valid transaction but the seller is not, so they do not receive the cash! This sort of thing is not
```

Author: Done

{{< /note >}}

#### Two Phase Finality

> From Corda 4.11 on, the finality protocol has changed to improve resilience and recoverability.

Reviewer: Release Note?

Author: Removed
